model/task_queue_service.go

```diff
@@ -0,0 +1,205 @@
+			if _, ok = units[id]; !ok {
+				units[id] = schedulableUnit{
+					id:       id,
+					maxHosts: item.GroupMaxHosts,
+					tasks:    []TaskQueueItem{item},
+				}
+				order = append(order, id)
+			} else {
+				unit = units[id]
+				unit.tasks = append(unit.tasks, item)
+				units[id] = unit
+			}
+		}
+	}
+	t := &taskDispatchService{
+		distroID: distroID,
+		order:    order,
+		units:    units,
+	}
+	return t
+}
+
+// FindNextTask returns the next dispatchable task in the queue.
+func (t *taskDispatchService) FindNextTask(spec TaskSpec) *TaskQueueItem {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	// If units is empty, the queue has been emptied. Reset order as an optimization so that the
+	// service no longer needs to iterate over it and check each item against the map.
+	if len(t.units) == 0 {
+		t.order = []string{}
+		return nil
+	}
+
+	var unit schedulableUnit
+	var ok bool
+	var next *TaskQueueItem
+
+	if spec.Group != "" {
+		unit, ok = t.units[compositeGroupId(spec.Group, spec.BuildVariant, spec.Version)]
+		if ok {
+			if next = t.nextTaskGroupTask(unit); next != nil {
+				return next
+			}
+		}
+		// If the task group is not present in the task group map, it has been dispatched.
+		// Fall through to getting a task not in that group.
+	}
+	var numHosts int
+	var err error
+	for _, schedulableUnitID := range t.order {
+		unit, ok = t.units[schedulableUnitID]
+		if !ok {
+			continue
+		}
+		// If maxHosts is not set, this is not a task group.
+		if unit.maxHosts == 0 {
+			delete(t.units, schedulableUnitID)
+			return &unit.tasks[0]
+		}
+		if unit.runningHosts < unit.maxHosts {
+			// TODO: For a multi-host task group, it's not possible to correctly
+			// dispatch tasks based on number of hosts running tasks in the task group
+			// without a transaction. When we use a driver that supports transactions,
+			// we likely want to rewrite this code to use transactions. Currently it's
+			// possible to dispatch a multi-host task group to more tasks than max
+			// hosts. Before transactions are available, it might be possible to do this
+			// better by performing this query again, just before returning from this
+			// function.
+			numHosts, err = host.NumHostsByTaskSpec(spec.Group, spec.BuildVariant, spec.ProjectID, spec.Version)
+			if err != nil {
+				grip.Error(message.WrapError(err, message.Fields{
+					"message": "problem running NumHostsByTaskSpec query",
+					"group":   spec.Group,
+					"variant": spec.BuildVariant,
+					"project": spec.ProjectID,
+					"version": spec.Version,
+				}))
+				return nil
+			}
+			unit.runningHosts = numHosts
+			t.units[schedulableUnitID] = unit
+			if unit.runningHosts < unit.maxHosts {
+				if next = t.nextTaskGroupTask(unit); next != nil {
+					return next
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func compositeGroupId(group, variant, version string) string {
+	return fmt.Sprintf("%s-%s-%s", group, variant, version)
```

Comment:
there has to be a function that does this canonically somewhere.

Reply:
There's not. There is never a time when we are appending variant and version to group. This is purely to create a key for the queue.